jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/profiles/ProfileWatcher.kt

```diff
@@ -4,148 +4,63 @@
 package software.aws.toolkits.jetbrains.core.credentials.profiles
 
 import com.intellij.openapi.Disposable
-import com.intellij.util.io.isFile
+import com.intellij.openapi.vfs.AsyncFileListener
+import com.intellij.openapi.vfs.LocalFileSystem
+import com.intellij.openapi.vfs.VfsUtilCore
+import com.intellij.openapi.vfs.VirtualFileManager
+import com.intellij.openapi.vfs.newvfs.events.VFileEvent
 import software.amazon.awssdk.profiles.ProfileFileLocation
-import software.aws.toolkits.core.utils.debug
 import software.aws.toolkits.core.utils.getLogger
-import software.aws.toolkits.core.utils.info
 import software.aws.toolkits.core.utils.tryOrNull
-import software.aws.toolkits.core.utils.warn
-import java.nio.file.FileSystems
-import java.nio.file.Files
-import java.nio.file.Path
-import java.nio.file.StandardWatchEventKinds
-import java.nio.file.WatchKey
 import java.util.concurrent.ConcurrentHashMap
-import java.util.concurrent.Executors
-import java.util.concurrent.Future
 
-class ProfileWatcher : Disposable {
+@Suppress("MissingRecentApi") // 2019.2 is 192.5728.98 TODO: Remove warning FIX_WHEN_MIN_IS_193
+class ProfileWatcher : AsyncFileListener, Disposable {
     private val listeners = ConcurrentHashMap.newKeySet<ProfileChangeListener>()
-    private val watchService = FileSystems.getDefault().newWatchService()
-    private val watchKeys: MutableMap<WatchKey, Path> = mutableMapOf()
-    private val watchLocations = setOf(
-        ProfileFileLocation.configurationFilePath().toAbsolutePath(),
-        ProfileFileLocation.credentialsFilePath().toAbsolutePath()
+    private val watchRoots = mutableSetOf<LocalFileSystem.WatchRequest>()
+
+    private val watchLocationsStrings = setOf(
+        ProfileFileLocation.configurationFilePath().toAbsolutePath().toString(),
+        ProfileFileLocation.credentialsFilePath().toAbsolutePath().toString()
     )
-    private val executor = Executors.newSingleThreadExecutor {
-        Thread(it).also { newThread ->
-            newThread.name = "AWSProfileWatcher"
-            newThread.isDaemon = true
+
+    override fun prepareChange(events: MutableList<out VFileEvent>): AsyncFileListener.ChangeApplier? {
+        val relevantChanges = events.filter { VfsUtilCore.isUnder(it.path, watchLocationsStrings) }.toList()
```

Comment:
since you just throw `relevantChanges` away after the first if block - why couldn't you do a `any()` match instead of a `filter` then a `toList` ?

Reply:
Good call out, will tweak this.